src/vs/workbench/api/browser/mainThreadChatCodeMapper.ts

```diff
@@ -56,9 +59,35 @@ export class MainThreadChatCodemapper extends Disposable implements MainThreadCo
 	$handleProgress(requestId: string, data: ICodeMapperProgressDto): Promise<void> {
 		const response = this._responseMap.get(requestId);
 		if (response) {
+			const edits = data.edits;
 			const resource = URI.revive(data.uri);
-			response.textEdit(resource, data.edits);
+			if (!edits.length) {
+				response.textEdit(resource, []);
+			} else if (areTextEdits(edits)) {
+				response.textEdit(resource, edits);
+			} else {
+				const cellEdits: ICellEditOperation[] = [];
+				edits.forEach(dto => {
+					if (dto.editType === CellEditType.Replace) {
+						cellEdits.push({
+							editType: dto.editType,
+							index: dto.index,
+							count: dto.count,
+							cells: dto.cells.map(NotebookDto.fromNotebookCellDataDto)
+						});
+					}
+				});
+				response.notebookEdit(resource, cellEdits);
+			}
 		}
 		return Promise.resolve();
 	}
 }
+
+function areTextEdits(edits: ICodeMapperProgressDto['edits']): edits is TextEdit[] {
+	if (edits.some(e => 'range' in e && 'text' in e)) {
```

Comment:
using `'string' in obj` isn't great, rename and reference search don't find them. It's better to cast `e` to `TextEdit` and then to probe. Maybe even have `TextEdit.is(...)` (there already exists this class) and only have the array-check here